crates/graphql/src/main.rs

```diff
@@ -35,38 +35,47 @@ fn graphiql(uri: String) -> impl Fn() -> HttpResponse + Clone {
     }
 }
 
-async fn graphql(
-    st: web::Data<Arc<Schema>>,
-    data: web::Json<GraphQLRequest>,
-) -> Result<HttpResponse, Error> {
-    let json = web::block(move || {
-        let res = data.execute_sync(&st, &());
-        serde_json::to_string(&res)
-    })
-    .await?;
-    Ok(HttpResponse::Ok()
-        .content_type("application/json")
-        .body(json))
+fn graphql(
+    db_pool: &Arc<Pool>,
+) -> impl Fn(
+    web::Data<Arc<Schema>>,
+    web::Json<GraphQLRequest>,
+) -> Pin<Box<dyn Future<Output = Result<HttpResponse, Error>> + Send>>
++ Clone {
+    let db_pool = Arc::clone(db_pool);
+
+    move |st, data| {
+        let db_connection = db_pool.get().unwrap();
+
+        Box::pin(async move {
+            let ctx = AppContext::new(db_connection);
+            let res = data.execute(&st, &ctx).await;
+
+            let json = serde_json::to_string(&res)?;
+
+            Ok(HttpResponse::Ok()
+                .content_type("application/json")
+                .body(json))
+        })
+    }
 }
 
 fn main() {
-    use indexer_core::prelude::*;
-
     indexer_core::run(|| {
         let Opts {
             server: ServerOpts { port },
         } = Opts::parse();
 
-        let db_conn =
-            db::connect(db::ConnectMode::Read).context("Failed to connect to Postgres")?;
+        let db_pool =
+            Arc::new(db::connect(db::ConnectMode::Read).context("Failed to connect to Postgres")?);
 
         let mut addr: SocketAddr = "0.0.0.0:3000".parse().unwrap();
         addr.set_port(port);
         info!("Listening on {}", addr);
 
         let graphiql_uri = format!("http://{}", addr);
 
-        let schema = std::sync::Arc::new(schema::create(db_conn));
+        let schema = Arc::new(schema::create(db_pool.clone()));
```

Comment:
Note to refactor attaching db_pool to the query root. Instead start reading the connection produced for the request.